### ISSUE TITLE
mixed_precision_graph_rewrite behaviours

### ISSUE THREAD

**Issue by don-tpanic** [tensorflow/tensorflow#33435]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.3
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.0
- Python version: 3.7
- Bazel version (if compiling from source): 
- GCC/Compiler version (if compiling from source): 7.4.0
- CUDA/cuDNN version: V10.0.130
- GPU model and memory: GeForce RTX 2080ti, 10G

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
See the results below. Basically, there no clear evidence mix precision speeds up the training in this case. 
**Describe the expected behavior**

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
```
import os
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"]= '3'
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

import tensorflow
import tensorflow.keras as keras
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, Dropout, Flatten, Input
from tensorflow.keras.layers import Conv2D, MaxPooling2D
from tensorflow.keras import backend as K
import time


optimizer1 = keras.optimizers.Adadelta()
optimizer2 = tensorflow.train.experimental.enable_mixed_precision_graph_rewrite(optimizer1)
opts_dict = {'fp32': optimizer1, 'mix': optimizer2}
batch_sizes = [256, 512, 1024, 2048, 4096, 8192]
num_classes = 10
epochs = 30

dataset = 'mnist'

for batch_size in batch_sizes:
    for precision in opts_dict:

        optimizer = opts_dict[precision]
        start_time = time.time()

        # the data, split between train and test sets
        if dataset == 'mnist':
            (x_train, y_train), (x_test, y_test) = mnist.load_data()
            if K.image_data_format() == 'channels_first':
                x_train = x_train.reshape(x_train.shape[0], 1, 28, 28)
                x_test = x_test.reshape(x_test.shape[0], 1, 28, 28)
                input_shape = (1, 28, 28)
            else:
                x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
                x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)
                input_shape = (28, 28, 1)

        if dataset == 'cifar10':
            (x_train, y_train), (x_test, y_test) = cifar10.load_data()
            input_shape = x_train.shape[1:]


        x_train = x_train.astype('float32')
        x_test = x_test.astype('float32')
        x_train /= 255
        x_test /= 255

        y_train = keras.utils.to_categorical(y_train, num_classes)
        y_test = keras.utils.to_categorical(y_test, num_classes)

        input = Input(shape=input_shape)
        x = Conv2D(32, kernel_size=(3, 3),
                         activation='relu',
                         input_shape=input_shape)(input)
        x = Conv2D(64, (3, 3), activation='relu')(x)
        x = MaxPooling2D(pool_size=(2, 2))(x)
        x = Dropout(0.25)(x)
        x = Flatten()(x)
        x = Dense(128, activation='relu')(x)
        x = Dropout(0.5)(x)
        x = Dense(num_classes, activation='softmax')(x)
        model = Model(inputs=input, outputs=x)
        model.compile(loss=keras.losses.categorical_crossentropy,
                      optimizer=optimizer,
                      metrics=['accuracy'])

        model.fit(x_train, y_train,
                  batch_size=batch_size,
                  epochs=epochs,
                  verbose=0,
                  validation_data=(x_test, y_test))
        score = model.evaluate(x_test, y_test, verbose=0)
        print('Batch size = %s' % batch_size)
        print('Precision = %s' % precision)
        print('Test loss:', score[0])
        print('Test accuracy:', score[1])
        dur = time.time() - start_time
        print('Run time = %s s' % dur)
        print('================================')


```
**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

Results on mnist:
![Screenshot from 2019-10-17 13-58-01](https://user-images.githubusercontent.com/32969920/67010775-33941d80-f0e6-11e9-944a-fccb1ec7b933.png)

I also tried running on cifar10:
![Screenshot from 2019-10-17 13-56-01](https://user-images.githubusercontent.com/32969920/67010652-f6c82680-f0e5-11e9-930c-e5f58bf7e0b7.png)

**Comment by reedwm** [comment]
You are running with mixed precision even for the fp32 runs. Mixed precision is enabled when you call the line:

```
optimizer2 = tensorflow.train.experimental.enable_mixed_precision_graph_rewrite(optimizer1)
```

Even if you do not use `optimizer2`, mixed precision is still being used. `optimizer2` will additionally use a technique called loss scaling to avoid numeric underflow, but mixed precision will still be used without it. 

The easiest way to fix this is running the Python script twice: once where you call `enable_mixed_precision_graph_rewrite` and once where you do not. Typically you do not want to run with both fp32 and mixed precision in a single process. Alternatively, you can start by running the mixed precision benchmarks, then call `tf.train.experimental.disable_mixed_precision_graph_rewrite()` before doing the fp32 benchmarks.

**Comment by superhg2012** [comment]
Hi,  I got these log infos after start training my model on T4 GPU card

`2019-11-07 16:51:44.992350: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1767] Running auto_mixed_precision graph optimizer
2019-11-07 16:51:44.997218: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1241] No whitelist ops found, nothing to do
2019-11-07 16:51:50.326678: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1767] Running auto_mixed_precision graph optimizer
2019-11-07 16:51:50.434888: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1723] Converted 1299/8694 nodes to float16 precision using 73 cast(s) to float16 (excluding Const and Variable casts)
2019-11-07 16:51:56.493836: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1767] Running auto_mixed_precision graph optimizer
2019-11-07 16:51:56.494237: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1241] No whitelist ops found, nothing to do
2019-11-07 16:51:56.957105: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1767] Running auto_mixed_precision graph optimizer
2019-11-07 16:51:56.957657: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1241] No whitelist ops found, nothing to do`

It seems nothing is wrong.
Then I follow the steps below to do inference 
1 load the checkpoint with tf. train.Saver() and restored the graph
2  export to savedmodel using saved_model_builder
3  deploying savedmodel files to tf-serving

 Everything is ok. But the inference performance is not impoved...
 It that ok?

**Comment by reedwm** [comment]
@superhg2012, `enable_mixed_precision_graph_rewrite` will have no effect on SavedModels when served with TensorFlow Serving. The reason is that the function causes TensorFlow to rewrite the graph behind the scenes to use mixed precision, but the SavedModel will still be in float32. 

In general, `enable_mixed_precision_graph_rewrite` is designed to be a training API (although it will work in inference if you do not export your model). The reason is that for training, more work must be done to ensure certain parts of the model stay in float32, while during inference, almost everything can be done in float16.

@suharshs, any pointers to using float16 for inference?

**Comment by superhg2012** [comment]
@reedwm thanks a lot for kindly explaination. if I want to run inference with tf.float16 dtype.  savedmodel can not be used. Do you mean I can only run inference with  the checkpoint after training?

**Comment by reedwm** [comment]
Yes, if you want to run inference in float16 with `enable_mixed_precision_graph_rewrite`, you must use the checkpoint after training. You can still save a SavedModel, but the SavedModel will not use mixed precision.

There might be an alternative way of using float16 with SavedModel. @christisg @netfs, do either of you know if you can choose to run a grappler pass when serving a SavedModel? Or use some other mechanism to make the model float16?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 44 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
We are closing this issue for now due to lack of activity. Please comment if this is still an issue for you. Thanks!

**Comment by lxl910915** [comment]
> @superhg2012, `enable_mixed_precision_graph_rewrite` will have no effect on SavedModels when served with TensorFlow Serving. The reason is that the function causes TensorFlow to rewrite the graph behind the scenes to use mixed precision, but the SavedModel will still be in float32.
> 
> In general, `enable_mixed_precision_graph_rewrite` is designed to be a training API (although it will work in inference if you do not export your model). The reason is that for training, more work must be done to ensure certain parts of the model stay in float32, while during inference, almost everything can be done in float16.
> 
> @suharshs, any pointers to using float16 for inference?

We also want to known is there any pointers to using float16 for inference?

**Comment by lxl910915** [comment]
set --platform_config_file=platform_config_file.cfg to enable float16 for inference.
```
platform_configs {
  key: "tensorflow"
  value {
    source_adapter_config {
      [type.googleapis.com/tensorflow.serving.SavedModelBundleSourceAdapterConfig] {
        legacy_config {
          #batching_parameters {
          #  max_batch_size { value: 100 }
          #  batch_timeout_micros { value: 200000000 }
          #  max_enqueued_batches { value: 10000 }
          #  num_batch_threads { value: 36 }  
          #}
          session_config {
            allow_soft_placement: true
            gpu_options {
              per_process_gpu_memory_fraction: 0.9
              allow_growth: true
            }
            graph_options {
              rewrite_options {
                auto_mixed_precision: 1 # DEFAULT = 0; ON = 1; OFF = 2;
              }
            }
          }
        }
      }
    }
  }
}

```
